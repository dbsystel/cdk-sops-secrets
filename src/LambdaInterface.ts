/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export interface SopsSyncResourcePropertys {
  ResourceType: "SECRET" | "SECRET_RAW" | "SECRET_BINARY" | "PARAMETER_MULTI" | "PARAMETER";
  Format: "json" | "yaml" | "dotenv" | "binary";
  Target: string;
  EncryptionKey?: string;
  SopsS3File?: SopsS3File;
  SopsInline?: SopsInline;
  FlattenSeparator?: string;
  ServiceToken?: string;
}
export interface SopsS3File {
  Bucket: string;
  Key: string;
}
export interface SopsInline {
  Content: string;
  Hash: string;
}
