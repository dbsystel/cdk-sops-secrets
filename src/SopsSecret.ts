import {
  AddToResourcePolicyResult,
  Grant,
  IGrantable,
  PolicyStatement,
} from 'aws-cdk-lib/aws-iam';
import { IKey } from 'aws-cdk-lib/aws-kms';
import {
  ISecret,
  ISecretAttachmentTarget,
  RotationSchedule,
  RotationScheduleOptions,
  Secret,
  SecretProps,
} from 'aws-cdk-lib/aws-secretsmanager';
import {
  RemovalPolicy,
  ResourceEnvironment,
  SecretValue,
  Stack,
} from 'aws-cdk-lib/core';
import { Construct } from 'constructs';
import {
  CreationType,
  ResourceType,
  SopsSync,
  SopsSyncOptions,
} from './SopsSync';

/**
 * The configuration options of the SopsSecret
 */
export interface SopsSecretProps extends SecretProps, SopsSyncOptions {}

/**
 * A drop in replacement for the normal Secret, that is populated with the encrypted
 * content of the given sops file.
 */
export class SopsSecret extends Construct implements ISecret {
  private readonly secret: Secret;
  readonly encryptionKey?: IKey | undefined;
  readonly secretArn: string;
  readonly secretFullArn?: string | undefined;
  readonly secretName: string;
  readonly stack: Stack;
  readonly env: ResourceEnvironment;

  readonly sync: SopsSync;
  public constructor(scope: Construct, id: string, props: SopsSecretProps) {
    super(scope, id);
    this.secret = new Secret(this, 'Resource', props as SecretProps);

    // Fullfill secret Interface
    this.encryptionKey = this.secret.encryptionKey;
    this.secretArn = this.secret.secretArn;
    this.secretName = this.secret.secretName;
    this.stack = Stack.of(scope);
    this.env = {
      account: this.stack.account,
      region: this.stack.region,
    };

    this.sync = new SopsSync(this, 'SopsSync', {
      secret: this.secret,
      resourceType: ResourceType.SECRET,
      creationType: CreationType.SINGLE,
      flattenSeparator: '.',
      ...(props as SopsSyncOptions),
    });
  }

  public currentVersionId(): string {
    return this.sync.versionId;
  }

  public grantRead(grantee: IGrantable, versionStages?: string[]): Grant {
    return this.secret.grantRead(grantee, versionStages);
  }
  public grantWrite(_grantee: IGrantable): Grant {
    throw new Error(
      `Method grantWrite(...) not allowed as this secret is managed by SopsSync`,
    );
  }
  public addRotationSchedule(
    id: string,
    options: RotationScheduleOptions,
  ): RotationSchedule {
    throw new Error(
      `Method addRotationSchedule('${id}', ${JSON.stringify(
        options,
      )}) not allowed as this secret is managed by SopsSync`,
    );
  }
  public addToResourcePolicy(
    statement: PolicyStatement,
  ): AddToResourcePolicyResult {
    return this.secret.addToResourcePolicy(statement);
  }
  public denyAccountRootDelete(): void {
    return this.secret.denyAccountRootDelete();
  }
  public attach(target: ISecretAttachmentTarget): ISecret {
    return this.secret.attach(target);
  }
  public applyRemovalPolicy(policy: RemovalPolicy): void {
    return this.secret.applyRemovalPolicy(policy);
  }

  public secretValueFromJson(jsonField: string) {
    return SecretValue.secretsManager(this.secretArn, {
      jsonField,
      versionId: this.sync.versionId,
    });
  }

  public get secretValue(): SecretValue {
    return this.secretValueFromJson('');
  }
}
