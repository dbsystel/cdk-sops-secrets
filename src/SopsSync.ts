import * as fs from 'fs';
import * as path from 'path';
import {
  SecretValue,
  Duration,
  Lazy,
  Stack,
  Annotations,
  CustomResource,
  FileSystem,
} from 'aws-cdk-lib';
import { IGrantable, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { IKey, Key } from 'aws-cdk-lib/aws-kms';
import { SingletonFunction, Code, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Asset } from 'aws-cdk-lib/aws-s3-assets';
import { ISecret } from 'aws-cdk-lib/aws-secretsmanager';
import { Construct } from 'constructs';

export enum UploadType {
  /**
   * Pass the secret data inline (base64 encoded and compressed)
   */
  INLINE = 'INLINE',
  /**
   * Uplaod the secert data as asset
   */
  ASSET = 'ASSET',
}

export enum CreationType {
  /**
   * Create or update a single secret/parameter
   */
  SINGLE = 'SINGLE',
  /**
   * Create or update a multiple secrets/parameters by flattening the SOPS file
   */
  MULTI = 'MULTI',
}

export enum ResourceType {
  SECRET = 'SECRET',
  PARAMETER = 'PARAMETER',
}

/**
 * Configuration options for the SopsSync
 */
export interface SopsSyncOptions {
  /**
   * The custom resource provider to use. If you don't specify any, a new
   * provider will be created - or if already exists within this stack - reused.
   *
   * @default - A new singleton provider will be created
   */
  readonly sopsProvider?: SopsSyncProvider;

  /**
   * The filepath to the sops file
   */
  readonly sopsFilePath?: string;

  /**
   * If you want to pass the sops file via s3, you can specify the bucket
   * you can use cfn parameter here
   * Both, sopsS3Bucket and sopsS3Key have to be specified
   */
  readonly sopsS3Bucket?: string;

  /**
   * If you want to pass the sops file via s3, you can specify the key inside the bucket
   * you can use cfn parameter here
   * Both, sopsS3Bucket and sopsS3Key have to be specified
   */
  readonly sopsS3Key?: string;

  /**
   * How should the secret be passed to the CustomResource?
   * @default INLINE
   */
  readonly uploadType?: UploadType;

  /**
   * The format of the sops file.
   *
   * @default - The fileformat will be derived from the file ending
   */
  readonly sopsFileFormat?: undefined | 'json' | 'yaml' | 'dotenv' | 'binary';

  /**
   * The kmsKey used to encrypt the sops file. Encrypt permissions
   * will be granted to the custom resource provider.
   *
   * @default - The key will be derived from the sops file
   */
  readonly sopsKmsKey?: IKey[];

  /**
   * The age key that should be used for encryption.
   */
  readonly sopsAgeKey?: SecretValue;

  /**
   * Should the encrypted sops value should be converted to JSON?
   * Only JSON can be handled by cloud formations dynamic references.
   *
   * @default true
   */
  readonly convertToJSON?: boolean;

  /**
   * Should the structure be flattened? The result will be a flat structure and all
   * object keys will be replaced with the full jsonpath as key.
   * This is usefull for dynamic references, as those don't support nested objects.
   *
   * @default true
   */
  readonly flatten?: boolean;

  /**
   * If the structure should be flattened use the provided separator between keys.
   *
   * @default '.'
   */
  readonly flattenSeparator?: string;

  /**
   * Add this prefix to parameter names.
   */
  readonly parameterKeyPrefix?: string;

  /**
   * Shall all values be flattened? This is usefull for dynamic references, as there
   * are lookup errors for certain float types
   */
  readonly stringifyValues?: boolean;

  readonly creationType?: CreationType;
  readonly resourceType?: ResourceType;
}

/**
 * The configuration options extended by the target Secret
 */
export interface SopsSyncProps extends SopsSyncOptions {
  /**
   * The secret that will be populated with the encrypted sops file content.
   */
  readonly secret?: ISecret;

  /**
   * The parameter name. If set this creates an encrypted SSM Parameter instead of a secret.
   */
  readonly parameterName?: string;

  /**
   * The parameter name. If set this creates an encrypted SSM Parameter instead of a secret.
   */
  readonly parameterNames?: string[];

  /**
   * The encryption key used for encrypting the ssm parameter if `parameterName` is set.
   */
  readonly encryptionKey?: IKey;
}

export class SopsSyncProvider extends SingletonFunction implements IGrantable {
  private sopsAgeKeys: SecretValue[];

  constructor(scope: Construct, id?: string) {
    super(scope, id ?? 'SopsSyncProvider', {
      code: Code.fromAsset(
        scope.node.tryGetContext('sops_sync_provider_asset_path') ||
          path.join(__dirname, '../assets/cdk-sops-lambda.zip'),
      ),
      runtime: Runtime.PROVIDED_AL2,
      handler: 'bootstrap',
      uuid: 'SopsSyncProvider',
      timeout: Duration.seconds(60),
      environment: {
        SOPS_AGE_KEY: Lazy.string({
          produce: () =>
            (this.sopsAgeKeys.map((secret) => secret.toString()) ?? []).join(
              '\n',
            ),
        }),
      },
    });
    this.sopsAgeKeys = [];
  }

  public addAgeKey(key: SecretValue) {
    this.sopsAgeKeys.push(key);
  }
}

/**
 * The custom resource, that is syncing the content from a sops file to a secret.
 */
export class SopsSync extends Construct {
  /**
   * The current versionId of the secret populated via this resource
   */
  readonly versionId: string;

  /**
   * Was the format converted to json?
   */
  readonly converToJSON: boolean;

  /**
   * Was the structure flattened?
   */
  readonly flatten: boolean;

  /**
   * Were the values stringified?
   */
  readonly stringifiedValues: boolean;

  constructor(scope: Construct, id: string, props: SopsSyncProps) {
    super(scope, id);

    this.converToJSON = props.convertToJSON ?? true;
    this.flatten = props.flatten ?? true;
    this.stringifiedValues = props.stringifyValues ?? true;

    const provider = props.sopsProvider ?? new SopsSyncProvider(scope);

    let uploadType = props.uploadType ?? UploadType.INLINE;
    let sopsFileFormat: 'json' | 'yaml' | 'dotenv' | 'binary' | undefined =
      props.sopsFileFormat;
    let sopsAsset: Asset | undefined = undefined;
    let sopsInline: { Content: string; Hash: string } | undefined = undefined;
    let sopsS3File: { Bucket: string; Key: string } | undefined = undefined;

    if (
      props.sopsFilePath !== undefined &&
      (props.sopsS3Bucket !== undefined || props.sopsS3Key !== undefined)
    ) {
      throw new Error(
        'You can either specify sopsFilePath or sopsS3Bucket and sopsS3Key!',
      );
    }

    if (props.sopsFilePath !== undefined) {
      const _sopsFileFormat =
        props.sopsFileFormat ?? props.sopsFilePath.split('.').pop();
      switch (_sopsFileFormat) {
        case 'json': {
          sopsFileFormat = 'json';
          break;
        }
        case 'yaml': {
          sopsFileFormat = 'yaml';
          break;
        }
        case 'yml': {
          sopsFileFormat = 'yaml';
          break;
        }
        case 'dotenv': {
          sopsFileFormat = 'dotenv';
          break;
        }
        case 'env': {
          sopsFileFormat = 'dotenv';
          break;
        }
        case 'binary': {
          sopsFileFormat = 'binary';
          break;
        }
        default: {
          throw new Error(`Unsupported sopsFileFormat ${_sopsFileFormat}`);
        }
      }

      if (!fs.existsSync(props.sopsFilePath)) {
        throw new Error(`File ${props.sopsFilePath} does not exist!`);
      }

      switch (uploadType) {
        case UploadType.INLINE: {
          sopsInline = {
            Content: fs.readFileSync(props.sopsFilePath).toString('base64'),
            // We calculate the hash the same way as it would be done by new Asset(..) - so we can ensure stable version names even if switching from INLINE to ASSET and viceversa.
            Hash: FileSystem.fingerprint(props.sopsFilePath),
          };
          break;
        }
        case UploadType.ASSET: {
          sopsAsset = new Asset(this, 'Asset', {
            path: props.sopsFilePath,
          });
          sopsS3File = {
            Bucket: sopsAsset.bucket.bucketName,
            Key: sopsAsset.s3ObjectKey,
          };
          break;
        }
      }

      if (provider.role !== undefined) {
        if (props.sopsKmsKey !== undefined) {
          props.sopsKmsKey.forEach((key) => key.grantDecrypt(provider.role!));
        }
        const fileContent = fs.readFileSync(props.sopsFilePath);
        // Handle keys
        const regexKey = /arn:aws:kms:[a-z0-9-]+:[\d]+:key\/[a-z0-9-]+/g;
        const resultsKey = fileContent.toString().match(regexKey);
        if (resultsKey !== undefined) {
          resultsKey?.forEach((result, index) =>
            Key.fromKeyArn(this, `SopsKey${index}`, result).grantDecrypt(
              provider.role!,
            ),
          );
        }
        const regexAlias = /arn:aws:kms:[a-z0-9-]+:[\d]+:alias\/[a-z0-9-]+/g;
        const resultsAlias = fileContent.toString().match(regexAlias);
        if (resultsAlias !== undefined) {
          resultsAlias?.forEach((result, index) =>
            Key.fromLookup(this, `SopsAlias${index}`, {
              aliasName: `alias/${result.split('/').slice(1).join('/')}`,
            }).grantDecrypt(provider.role!),
          );
        }
        if (props.secret) {
          props.secret.grantWrite(provider);
        }
        if (props.parameterName) {
          provider.addToRolePolicy(
            new PolicyStatement({
              actions: ['ssm:PutParameter'],
              resources: [
                `arn:aws:ssm:${Stack.of(this).region}:${
                  Stack.of(this).account
                }:parameter/${props.parameterName}`,
              ],
            }),
          );
          props.encryptionKey?.grantEncryptDecrypt(provider);
        }
        if (props.parameterNames) {
          provider.addToRolePolicy(
            new PolicyStatement({
              actions: ['ssm:PutParameter'],
              resources: props.parameterNames.map(
                (param) =>
                  `arn:aws:ssm:${Stack.of(this).region}:${
                    Stack.of(this).account
                  }:parameter${param.startsWith('/') ? param : `/${param}`}`,
              ),
            }),
          );
          props.encryptionKey?.grantEncryptDecrypt(provider);
        }
        if (sopsAsset !== undefined) {
          sopsAsset.bucket.grantRead(provider);
        }
        /**
         * fixes #234
         * If the kms key for secrets encryption is an IKey
         * there will be no permissions otherwise
         */
        if (
          props.secret?.encryptionKey !== undefined &&
          !(props.secret.encryptionKey instanceof Key)
        ) {
          props.secret.encryptionKey.grantEncryptDecrypt(provider);
        }
      } else {
        Annotations.of(this).addWarning(
          `Please ensure proper permissions for the passed lambda function:\n  - write Access to the secret\n  - encrypt with the sopsKmsKey${
            uploadType === UploadType.ASSET
              ? '\n  - download from asset bucket'
              : ''
          }`,
        );
      }
      if (props.sopsAgeKey !== undefined) {
        provider.addAgeKey(props.sopsAgeKey);
      }
    } else if (
      props.sopsS3Bucket !== undefined &&
      props.sopsS3Key !== undefined
    ) {
      sopsS3File = {
        Bucket: props.sopsS3Bucket,
        Key: props.sopsS3Key,
      };
      uploadType = UploadType.ASSET;
      Annotations.of(this).addWarning(
        'You have to manually add permissions to the sops provider to (permission to download file, to decrypt sops file)!',
      );
    } else {
      throw new Error(
        'You have to specify both sopsS3Bucket and sopsS3Key or neither!',
      );
    }

    if (sopsFileFormat === undefined) {
      throw new Error('You have to specify sopsFileFormat!');
    }

    const cr = new CustomResource(this, 'Resource', {
      serviceToken: provider.functionArn,
      resourceType: 'Custom::SopsSync',
      properties: {
        SecretARN: props.secret?.secretArn,
        SopsS3File: sopsS3File,
        SopsInline: sopsInline,
        ConvertToJSON: this.converToJSON,
        Flatten: this.flatten,
        FlattenSeparator: props.flattenSeparator ?? '.',
        ParameterKeyPrefix: props.parameterKeyPrefix,
        Format: sopsFileFormat,
        StringifiedValues: this.stringifiedValues,
        ParameterName: props.parameterName,
        EncryptionKey:
          props.secret !== undefined ? undefined : props.encryptionKey?.keyId,
        ResourceType: props.resourceType
          ? props.resourceType.toString()
          : ResourceType.SECRET.toString(),
        CreationType: props.creationType
          ? props.creationType.toString()
          : CreationType.SINGLE.toString(),
      },
    });
    this.versionId = cr.getAttString('VersionId');
  }
}
