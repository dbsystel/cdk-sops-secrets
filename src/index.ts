export * from './SopsSync';
export * from './SopsStringParameter';
export * from './SopsSecret';
export * from './MultiStringParameter';
