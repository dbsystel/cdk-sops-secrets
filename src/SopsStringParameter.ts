import { Grant, IGrantable } from 'aws-cdk-lib/aws-iam';
import { IKey } from 'aws-cdk-lib/aws-kms';
import {
  IStringParameter,
  StringParameter,
  StringParameterProps,
} from 'aws-cdk-lib/aws-ssm';
import { RemovalPolicy, ResourceEnvironment, Stack } from 'aws-cdk-lib/core';
import { Construct } from 'constructs';
import { SopsSync, SopsSyncOptions } from './SopsSync';

/**
 * The configuration options of the StringParameter
 */
export interface SopsStringParameterProps
  extends SopsSyncOptions,
    StringParameterProps {
  readonly encryptionKey: IKey;
}

/**
 * A drop in replacement for the normal String Parameter, that is populated with the encrypted
 * content of the given sops file.
 */
export class SopsStringParameter extends Construct implements IStringParameter {
  private readonly parameter: StringParameter;
  readonly sync: SopsSync;
  readonly encryptionKey: IKey;
  readonly stack: Stack;
  readonly env: ResourceEnvironment;
  readonly parameterArn: string;
  readonly parameterName: string;
  readonly parameterType: string;
  readonly stringValue: string;

  public constructor(
    scope: Construct,
    id: string,
    props: SopsStringParameterProps,
  ) {
    super(scope, id);

    this.encryptionKey = props.encryptionKey;
    this.stack = Stack.of(scope);
    this.env = {
      account: this.stack.account,
      region: this.stack.region,
    };

    this.parameter = new StringParameter(this, 'Resource', {
      ...(props as StringParameterProps),
      stringValue: ' ',
    });
    this.parameterArn = this.parameter.parameterArn;
    this.parameterName = this.parameter.parameterName;
    this.parameterType = this.parameter.parameterType;
    this.stringValue = this.parameter.stringValue;

    this.sync = new SopsSync(this, 'SopsSync', {
      encryptionKey: this.parameter.encryptionKey,
      parameterName: this.parameter.parameterName,
      ...(props as SopsSyncOptions),
    });
  }
  grantRead(grantee: IGrantable): Grant {
    return this.parameter.grantRead(grantee);
  }
  grantWrite(grantee: IGrantable): Grant {
    return this.parameter.grantWrite(grantee);
  }
  applyRemovalPolicy(policy: RemovalPolicy): void {
    this.parameter.applyRemovalPolicy(policy);
  }
}
