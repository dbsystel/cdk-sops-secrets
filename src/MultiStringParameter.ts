import * as fs from 'fs';
import { IKey } from 'aws-cdk-lib/aws-kms';
import { ParameterTier, StringParameter } from 'aws-cdk-lib/aws-ssm';
import { ResourceEnvironment, Stack } from 'aws-cdk-lib/core';
import { Construct } from 'constructs';
import * as YAML from 'yaml';
import { SopsStringParameterProps } from './SopsStringParameter';
import {
  CreationType,
  ResourceType,
  SopsSync,
  SopsSyncOptions,
} from './SopsSync';

interface JSONObject {
  [key: string]: any;
}

export interface MultiStringParameterProps extends SopsStringParameterProps {
  readonly keySeperator?: string;
  readonly keyPrefix?: string;
}

function flattenJSON(
  data: JSONObject,
  parentKey: string = '',
  result: JSONObject = {},
  keySeparator = '',
): JSONObject {
  for (let key in data) {
    if (data.hasOwnProperty(key)) {
      let newKey = parentKey ? `${parentKey}${keySeparator}${key}` : key;
      if (Array.isArray(data[key])) {
        data[key].forEach((item: JSONObject | null, index: any) => {
          let arrayKey = `${newKey}[${index}]`;
          if (typeof item === 'object' && item !== null) {
            flattenJSON(item, arrayKey, result, keySeparator);
          } else {
            result[arrayKey] = item;
          }
        });
      } else if (typeof data[key] === 'object' && data[key] !== null) {
        flattenJSON(data[key], newKey, result, keySeparator);
      } else {
        result[newKey] = data[key];
      }
    }
  }
  return result;
}

export class MultiStringParameter extends Construct {
  readonly sync: SopsSync;
  readonly encryptionKey: IKey;
  readonly stack: Stack;
  readonly env: ResourceEnvironment;
  readonly keyPrefix: string;
  readonly keySeparator: string;

  constructor(scope: Construct, id: string, props: MultiStringParameterProps) {
    super(scope, id);

    this.encryptionKey = props.encryptionKey;
    this.stack = Stack.of(scope);
    this.env = {
      account: this.stack.account,
      region: this.stack.region,
    };
    this.keyPrefix = props.keyPrefix ?? '/';
    this.keySeparator = props.keySeperator ?? '/';

    const keys = this.parseFile(props.sopsFilePath!, this.keySeparator)
      .filter((key) => !key.startsWith('sops'))
      .map((value) => `${this.keyPrefix}${value}`);

    keys.forEach((key) => {
      new StringParameter(this, 'Resource' + key, {
        parameterName: key,
        tier: ParameterTier.STANDARD,
        stringValue: ' ',
      });
    });

    this.sync = new SopsSync(this, 'SopsSync', {
      encryptionKey: this.encryptionKey,
      resourceType: ResourceType.PARAMETER,
      creationType: CreationType.MULTI,
      flatten: true,
      flattenSeparator: this.keySeparator,
      parameterKeyPrefix: this.keyPrefix,
      parameterNames: keys,
      ...(props as SopsSyncOptions),
    });
  }

  private parseFile(sopsFilePath: string, keySeparator: string): string[] {
    const _sopsFileFormat = sopsFilePath.split('.').pop();
    switch (_sopsFileFormat) {
      case 'json': {
        return Object.keys(
          flattenJSON(
            JSON.parse(fs.readFileSync(sopsFilePath, 'utf-8')),
            '',
            undefined,
            keySeparator,
          ),
        );
      }
      case 'yaml': {
        const content = fs.readFileSync(sopsFilePath, 'utf-8');
        const data = YAML.parse(content) as JSONObject;
        return Object.keys(flattenJSON(data, '', undefined, keySeparator));
      }
      default: {
        throw new Error(
          `Unsupported sopsFileFormat for multiple parameters: ${_sopsFileFormat}`,
        );
      }
    }
  }
}
