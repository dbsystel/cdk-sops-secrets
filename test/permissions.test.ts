import * as fs from 'fs';
import path from 'path';
import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { IRole, Role } from 'aws-cdk-lib/aws-iam';
import { Key } from 'aws-cdk-lib/aws-kms';
import { Permissions } from '../src/SopsSync';

describe('keysFromSopsContent', () => {
  let stack: Stack;

  beforeEach(() => {
    stack = new Stack();
  });

  test('returns an empty array when no keys are found', () => {
    const content = 'no keys here';
    const result = Permissions.keysFromSopsContent(stack, content);
    expect(result).toEqual([]);
  });

  test('returns an array of IKey objects when keys are found', () => {
    const content = fs
      .readFileSync(
        path.join(__dirname, '../test-secrets/yaml/sopsfile.enc-kms.yaml'),
      )
      .toString();
    const result = Permissions.keysFromSopsContent(stack, content);
    expect(result).toHaveLength(1);
    expect(result[0].keyArn).toBe(
      'arn:aws:kms:aws-region-1:123456789011:key/00000000-1234-4321-abcd-1234abcd12ab',
    );
    expect(result[0].keyId).toBe('00000000-1234-4321-abcd-1234abcd12ab');
  });

  test('returns multiple IKey objects when multiple keys are found', () => {
    const content = fs
      .readFileSync(
        path.join(__dirname, '../test-secrets/yaml/sopsfile.enc-multikms.yaml'),
      )
      .toString();
    const result = Permissions.keysFromSopsContent(stack, content);
    expect(result).toHaveLength(4);
    expect(result[0].keyArn).toBe(
      'arn:aws:kms:aws-region-1:123456789011:key/00000000-1234-4321-abcd-1234abcd12ab',
    );
    expect(result[0].keyId).toBe('00000000-1234-4321-abcd-1234abcd12ab');
    expect(result[1].keyArn).toBe(
      'arn:aws:kms:aws-region-1:123456789011:key/00000001-1234-4321-abcd-1234abcd12ab',
    );
    expect(result[1].keyId).toBe('00000001-1234-4321-abcd-1234abcd12ab');
    expect(result[2].keyArn).toBe(
      'arn:aws:kms:aws-region-1:123456789011:key/00000002-1234-4321-abcd-1234abcd12ab',
    );
    expect(result[2].keyId).toBe('00000002-1234-4321-abcd-1234abcd12ab');
    expect(result[3].keyArn).toBe(
      'arn:aws:kms:aws-region-1:123456789011:key/00000003-1234-4321-abcd-1234abcd12ab',
    );
    expect(result[3].keyId).toBe('00000003-1234-4321-abcd-1234abcd12ab');
  });
});

describe('keysFromSopsContentAlias', () => {
  let stack: Stack;

  beforeEach(() => {
    stack = new Stack();
  });

  test('returns an empty array when no keys are found', () => {
    const content = 'no keys here';
    const result = Permissions.keysFromSopsContentAlias(stack, content);
    expect(result).toEqual([]);
  });

  test('returns an array of IKey objects when keys are found', () => {
    jest
      .spyOn(Key, 'fromLookup')
      .mockReturnValue(
        Key.fromKeyArn(
          stack,
          'Key',
          'arn:aws:kms:aws-region-1:123456789011:key/00000000-1234-4321-abcd-1234abcd12ab',
        ),
      );

    const content = fs
      .readFileSync(
        path.join(
          __dirname,
          '../test-secrets/yaml/sopsfile.enc-kms-alias.yaml',
        ),
      )
      .toString();
    const result = Permissions.keysFromSopsContentAlias(stack, content);
    expect(result[0].keyArn).toBe(
      'arn:aws:kms:aws-region-1:123456789011:key/00000000-1234-4321-abcd-1234abcd12ab',
    );
    expect(result[0].keyId).toBe('00000000-1234-4321-abcd-1234abcd12ab');
  });
});

describe('parameters', () => {
  let stack: Stack;
  let role: IRole;

  beforeEach(() => {
    stack = new Stack();
    role = Role.fromRoleArn(
      stack,
      'TestRole',
      'arn:aws:iam::123456789012:role/test-role',
    );
  });

  function genParameters(prefix: string, count: number) {
    const params: string[] = [];
    for (let i = 0; i < count; i++) {
      params.push(`${prefix}${i}`);
    }
    return params;
  }

  test('1 parameter - full snapshot', () => {
    const params = genParameters('parameter', 1);
    Permissions.parameters(stack, params, role);
    expect(
      Template.fromStack(stack).findResources('AWS::IAM::ManagedPolicy'),
    ).toMatchSnapshot();
  });

  test('100 parameter - snapshot count', () => {
    const params = genParameters('parameter', 100);
    Permissions.parameters(stack, params, role);
    expect(
      Object.keys(
        Template.fromStack(stack).findResources('AWS::IAM::ManagedPolicy'),
      ),
    ).toMatchSnapshot();
  });

  test('1000 parameter - snapshot count', () => {
    const params = genParameters('parameter', 1000);
    Permissions.parameters(stack, params, role);
    expect(
      Object.keys(
        Template.fromStack(stack).findResources('AWS::IAM::ManagedPolicy'),
      ),
    ).toMatchSnapshot();
  });
});
