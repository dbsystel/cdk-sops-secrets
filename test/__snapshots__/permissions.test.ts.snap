// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parameters 1 parameter - full snapshot 1`] = `
Object {
  "SopsSecretParameterProviderManagedPolicyParameterAccess03C7CE5D3": Object {
    "Properties": Object {
      "Description": "Policy to grant parameter provider permissions to put parameter",
      "Path": "/",
      "PolicyDocument": Object {
        "Statement": Array [
          Object {
            "Action": "ssm:PutParameter",
            "Effect": "Allow",
            "Resource": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "arn:aws:ssm:",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":parameter/parameter0",
                ],
              ],
            },
          },
        ],
        "Version": "2012-10-17",
      },
    },
    "Type": "AWS::IAM::ManagedPolicy",
  },
}
`;

exports[`parameters 100 parameter - snapshot count 1`] = `
Array [
  "SopsSecretParameterProviderManagedPolicyParameterAccess03C7CE5D3",
  "SopsSecretParameterProviderManagedPolicyParameterAccess1E27103A2",
]
`;

exports[`parameters 1000 parameter - snapshot count 1`] = `
Array [
  "SopsSecretParameterProviderManagedPolicyParameterAccess03C7CE5D3",
  "SopsSecretParameterProviderManagedPolicyParameterAccess1E27103A2",
  "SopsSecretParameterProviderManagedPolicyParameterAccess204C56AEE",
  "SopsSecretParameterProviderManagedPolicyParameterAccess39C2C7238",
  "SopsSecretParameterProviderManagedPolicyParameterAccess4810168FC",
  "SopsSecretParameterProviderManagedPolicyParameterAccess5A4020F4D",
  "SopsSecretParameterProviderManagedPolicyParameterAccess6D18916A5",
  "SopsSecretParameterProviderManagedPolicyParameterAccess7C1D2D5DD",
  "SopsSecretParameterProviderManagedPolicyParameterAccess881B83B3B",
  "SopsSecretParameterProviderManagedPolicyParameterAccess9F6C18C5F",
  "SopsSecretParameterProviderManagedPolicyParameterAccess10B80D3BA8",
  "SopsSecretParameterProviderManagedPolicyParameterAccess118D826D5D",
]
`;
